java/src/main/native/src/RmmJni.cpp

```diff
@@ -365,6 +367,11 @@ JNIEXPORT void JNICALL Java_ai_rapids_cudf_Rmm_initializeInternal(
         Initialized_resource = rmm::mr::make_owning_wrapper<rmm::mr::arena_memory_resource>(
             std::make_shared<rmm::mr::cuda_memory_resource>(), pool_size, pool_limit);
       }
+    } else if (use_cuda_async_alloc) {
+      auto pool_limit = (max_pool_size > 0) ? thrust::optional<std::size_t>{max_pool_size} :
+                                              thrust::optional<std::size_t>{};
+      Initialized_resource = std::make_shared<rmm::mr::cuda_async_memory_resource>(
+            thrust::optional<std::size_t>{pool_size}, pool_limit);
```

Comment:
We're passing `pool_limit` as the `release_threshold` parameter.  Does that really act like a total pool limit?  The docs imply that it simply tries to release unused memory when the pool gets to that size, but it doesn't indicate that allocations will fail if it is unable to release any memory at that point.

Reply:
The `pool_limit` here has a different meaning than something like the `limiting_adaptor`. Allocations will not fail above this limit. Instead, this limits the amount of memory the cuda pool is allowed to hold on to.

Reply:
The caller thinks it is specifying a maximum total allocator limit to keep GPU memory utilization below a certain threshold.  If they didn't want any limit, they would set this to 0.  The caller doesn't have the concept of a "retain this much but let me go over if I need to", so I'm not sure mapping the original maximum allocator size to the release threshold makes sense.

Reply:
According to https://developer.nvidia.com/blog/using-cuda-stream-ordered-memory-allocator-part-1/:
> The pool threshold is just a hint. Memory in the pool can also be released implicitly by the CUDA driver to enable an  unrelated memory allocation request in the same process to succeed. For example, a call to cudaMalloc or cuMemCreate could cause CUDA to free unused memory from any memory pool associated with the device in the  same process to serve the request.

So it's not exactly the same as the max size for other pools, but it's close to what we want: retain as much memory as possible while leaving some reserve for other things (kernels, `cudaMalloc` calls). An allocation will eventually fail if there are no free blocks left to defragment.

Reply:
The problem is that we need to get an OOM before all memory is exhausted because it triggers us to spill.  If we're allowed to grow the pool and cache everything along the way, we'll eventually exhaust all of GPU memory and then some kernel launches will fail because there's insufficient free GPU memory.  The driver doesn't go through RMM to launch kernels, so we can't spill to help get that kernel launched.

We need this to be a hard limit, so I think we need to wrap this in a limit adaptor if the max pool size is specified.

Reply:
> we'll eventually exhaust all of GPU memory and then some kernel launches will fail because there's insufficient free GPU memory. The driver doesn't go through RMM to launch kernels, so we can't spill to help get that kernel launched.

I don't think this will happen. One of the unique parts of `cudaMallocAsync` is that even if your pool is holding on to 100% of GPU memory, a `cudaMalloc` call will force the pool to let go of some of it's memory.  We could never do something like this with RMM's pool implementation because we have no knowledge of other allocations, but since `cudaMallocAsync` and `cudaMalloc` are both going through the driver, it _can_ do stuff like that.

Reply:
I think we do need the hard limit in case there are multiple processes accessing the same GPU (e.g. python UDF).